presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java

```diff
@@ -0,0 +1,266 @@
+    {
+        this.context = requireNonNull(context, "context is null");
+        this.maxFilterPositionsCount = maxFilterPositionsCount;
+        this.maxFilterSizeInBytes = maxFilterSize.toBytes();
+
+        this.dynamicPredicateConsumer = requireNonNull(dynamicPredicateConsumer, "dynamicPredicateConsumer is null");
+        this.channels = requireNonNull(channels, "channels is null");
+
+        this.blockBuilders = new BlockBuilder[channels.size()];
+        this.valueSets = new TypedSet[channels.size()];
+        for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            Type type = channels.get(channelIndex).type;
+            this.blockBuilders[channelIndex] = type.createBlockBuilder(null, EXPECTED_BLOCK_BUILDER_SIZE);
+            this.valueSets[channelIndex] = new TypedSet(type, Optional.empty(), blockBuilders[channelIndex], EXPECTED_BLOCK_BUILDER_SIZE,
+                                                        String.format("DynamicFilterSourceOperator_%s_%d", planNodeId, channelIndex), maxFilterSize);
+        }
+        this.filterPositionsCount = 0;  // non-null values count
+        this.filterSizeInBytes = 0;  // non-null values size in bytes
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return context;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return current == null && !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(!finished, "DynamicFilterSourceOperator: addInput() may not be called after finish()");
+        current = page;
+        if (valueSets == null) {
+            return;  // the predicate became too large.
+        }
+
+        // Collect only the columns which are relevant for the JOIN.
+        try {
+            for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+                Block block = page.getBlock(channels.get(channelIndex).index);
+                TypedSet valueSet = valueSets[channelIndex];
+                filterSizeInBytes -= valueSet.getRetainedSizeInBytes();
+                filterPositionsCount -= valueSet.size();
+                for (int position = 0; position < block.getPositionCount(); ++position) {
+                    valueSet.add(block, position);
+                }
+                filterSizeInBytes += valueSet.getRetainedSizeInBytes();
+                filterPositionsCount += valueSet.size();
+            }
+            if (filterPositionsCount > maxFilterPositionsCount || filterSizeInBytes > maxFilterSizeInBytes) {
+                handleTooLargePredicate();
+            }
+        }
+        catch (PrestoException e) {
+            if (e.getErrorCode().equals(EXCEEDED_FUNCTION_MEMORY_LIMIT.toErrorCode())) {
```

Comment:
Wouldn't it be hard for users to figure out why dynamic filtering didn't apply with this approach ? How about we add an upper limit to DYNAMIC_FILTERING_MAX_PER_DRIVER_BYTE_SIZE in SystemSessionProperties ? There is already an upper bound of 1MB in FeatureConfig on `experimental.dynamic-filtering-max-per-driver-byte-size`

Reply:
Following @sopel39 [suggestion](https://github.com/prestosql/presto/pull/931#discussion_r319014865), I am using a new `TypedSet` c-tor, which allows setting `maxBlockMemoryInBytes` to the specified session property:
https://github.com/prestosql/presto/blob/5fbf6338955ec70c3bc147b9059ab5157c7dc6f5/presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java#L153-L158
IIUC, it should throw `PrestoException` only if a single block builder exceeds the specified `maxBlockMemoryInBytes`.
What do you think?

Reply:
I see, is the condition when exception is thrown essentially the same as `filterSizeInBytes > maxFilterSizeInBytes` ? I'm a bit confused by the presence of both the try/catch and the `filterSizeInBytes > maxFilterSizeInBytes` condition. If the TypedSet now already enforces maxFilterSizeInBytes by throwing exception, then maybe we get rid of the code around `filterSizeInBytes` ?

Reply:
Good point, added a comment to make this clearer.